core/webif.ml

```diff
@@ -160,14 +160,36 @@ struct
   let do_request ((valenv, _, _) as env) cgi_args run render_cont render_servercont_cont response_printer =
     let request = parse_request env cgi_args in
     let (>>=) f g = Lwt.bind f g in
+    (* We need to be a bit careful about what we respond here. If we are evaluating
+     * a ServerCont or EvalMain, that is fine -- but we need to construct a b64-encoded
+     * JSON object if we're responding to a ClientReturn or RemoteCall. *)
     Lwt.catch
       (fun () -> perform_request valenv run render_cont render_servercont_cont request )
-      (function
-       | Aborted r -> Lwt.return r
-       | Failure msg as e ->
-          prerr_endline msg;
-          Lwt.return ("text/html; charset=utf-8", error_page (Errors.format_exception_html e))
-       | exc -> Lwt.return ("text/html; charset=utf-8", error_page (Errors.format_exception_html exc)))
-    >>= fun (content_type, content) ->
-    response_printer [("Content-type", content_type)] content
+      (fun e ->
+        if (is_ajax_call cgi_args) then
```

Comment:
This function/ if-then-else in the middle of the >>= pipeline makes it more difficult to follow what's going on 0 roughly, do X, (if something then do Y1 else Y2); do Z.

Morally I feel like the middle (error handling) part should be two functions (that could maybe be hoisted earlier and called things like `handle_ajax_error` and `handle_regular_error`) and we choose which one based on is_ajax_call which does not depend on e.